File: anqa/harness/copilot.py
"""GitHub Copilot CLI disk adapter (``~/.copilot/session-store.db``).

Sessions are sqlite rows. Timeline is ``session-state/<id>/events.jsonl``.
"""

from __future__ import annotations

import shutil
import sqlite3
import tarfile
from collections.abc import Sequence
from pathlib import Path

from .. import event_types as et
from ..json_lines import json_lines
from ..models import JsonObject, SessionMeta, ToolInputBag, TraceEvent, json_mapping
from ..stamp import Stamp
from .ref import SessionRef
from .status import from_last

COPILOT_HARNESS_ID = "copilot"
_TURN_SIGNALS = frozenset(
    {
        "assistant.turn_start",
        "tool.execution_start",
        "subagent.started",
        "session.shutdown",
        "assistant.turn_end",
    }
)


def default_store_root() -> Path:
    """Host Copilot config tree (resolved at call time)."""
    return Path.home() / ".copilot"


def default_db_path() -> Path:
    """Host Copilot catalog database."""
    return default_store_root() / "session-store.db"


def _connect(db: Path) -> sqlite3.Connection:
    path = Path(db).expanduser()
    con = sqlite3.connect(f"file:{path}?mode=ro", uri=True)
    con.row_factory = sqlite3.Row
    return con


def _assert_readable(db: Path) -> Path:
    path = Path(db).expanduser()
    if not path.is_file():
        raise FileNotFoundError(f"copilot database not found: {path}")
    return path


def _db_from_ref(ref: SessionRef | Path | str, fallback: Path) -> tuple[Path, str]:
    if isinstance(ref, SessionRef):
        return Path(ref.locator), ref.session_id
    text = str(ref)
    from .ref import parse_session_ref_string

    parsed = parse_session_ref_string(text)
    if parsed is not None:
        return fallback, parsed[1]
    path = Path(text).expanduser()
    if path.is_file():
        return path, ""
    return fallback, path.name


def _state_dir(db: Path, session_id: str) -> Path:
    return Path(db).expanduser().resolve().parent / "session-state" / session_id


def _events_path(db: Path, session_id: str) -> Path:
    return _state_dir(db, session_id) / "events.jsonl"


def _session_row(con: sqlite3.Connection, session_id: str) -> sqlite3.Row | None:
    return con.execute(
        "SELECT id, cwd, repository, host_type, branch, summary, created_at, updated_at "
        "FROM sessions WHERE id = ?",
        (session_id,),
    ).fetchone()


def _list_session_rows(con: sqlite3.Connection) -> list[sqlite3.Row]:
    return list(
        con.execute(
            "SELECT id, cwd, repository, host_type, branch, summary, created_at, updated_at "
            "FROM sessions ORDER BY updated_at DESC"
        )
    )


def _last_turn_type(events: list[JsonObject]) -> str:
    last = ""
    for ev in events:
        typ = str(ev.get("type") or "").strip()
        if typ in _TURN_SIGNALS:
            last = typ
    return last


def _turn_outcome(events: list[JsonObject]) -> str:
    return from_last(_last_turn_type(events))


def _text_of(raw: object) -> str:
    if isinstance(raw, str):
        return raw
    if isinstance(raw, dict):
        return str(raw.get("content") or raw.get("text") or "").strip()
    return ""


def _timeline_for(events: list[JsonObject]) -> list[TraceEvent]:
    out: list[TraceEvent] = []
    for i, row in enumerate(events):
        ev = _event_from_row(i, row)
        if ev is not None:
            out.append(ev)
    return out


def _event_from_row(index: int, row: JsonObject) -> TraceEvent | None:
    typ = str(row.get("type") or "").strip()
    data = json_mapping(row.get("data"))
    ts = Stamp.epoch(row.get("timestamp"))
    eid = str(row.get("id") or "")
    if typ in {"session.start", "assistant.turn_start"}:
        return TraceEvent(
            index=index,
            event_type=et.TURN_STARTED,
            timestamp=ts,
            content="",
            update_index=index,
        )
    if typ == "assistant.turn_end":
        return TraceEvent(
            index=index,
            event_type=et.TURN_ENDED,
            timestamp=ts,
            content="",
            update_index=index,
        )
    if typ == "session.shutdown":
        return TraceEvent(
            index=index,
            event_type=et.TURN_COMPLETED,
            timestamp=ts,
            content=str(data.get("shutdownType") or "").strip(),
            update_index=index,
        )
    if typ == "user.message":
        return TraceEvent(
            index=index,
            event_type=et.USER_MESSAGE_CHUNK,
            timestamp=ts,
            content=_text_of(data.get("content")),
            update_index=index,
        )
    if typ == "assistant.message":
        return TraceEvent(
            index=index,
            event_type=et.AGENT_MESSAGE_CHUNK,
            timestamp=ts,
            content=_text_of(data.get("content")),
            update_index=index,
        )
    if typ == "tool.execution_start":
        args = json_mapping(data.get("arguments"))
        name = str(data.get("toolName") or "").strip()
        return TraceEvent(
            index=index,
            event_type=et.TOOL_CALL,
            timestamp=ts,
            tool_name=name,
            tool_call_id=str(data.get("toolCallId") or eid),
            content="",
            raw_input=ToolInputBag(args),
            update_index=index,
        )
    if typ == "tool.execution_complete":
        ok = data.get("success") is True
        return TraceEvent(
            index=index,
            event_type=et.TOOL_CALL_UPDATE,
            timestamp=ts,
            tool_call_id=str(data.get("toolCallId") or eid),
            content=_text_of(data.get("result")),
            is_error=not ok,
            update_index=index,
        )
    if typ == "subagent.started":
        name = str(data.get("agentName") or data.get("agentDisplayName") or "").strip()
        child = str(row.get("agentId") or data.get("agentId") or "").strip()
        return TraceEvent(
            index=index,
            event_type=et.SUBAGENT_SPAWNED,
            timestamp=ts,
            content=name,
            tool_call_id=str(data.get("toolCallId") or ""),
            raw_input=ToolInputBag(
                {
                    "child_session_id": child,
                    "subagent_id": child,
                    "subagent_type": str(data.get("agentType") or ""),
                    "description": str(data.get("agentDescription") or name),
                }
            ),
            update_index=index,
        )
    if typ == "subagent.completed":
        child = str(row.get("agentId") or data.get("agentId") or "").strip()
        cancelled = data.get("cancelled") is True
        return TraceEvent(
            index=index,
            event_type=et.SUBAGENT_FINISHED,
            timestamp=ts,
            content=str(data.get("agentName") or "").strip(),
            tool_call_id=str(data.get("toolCallId") or ""),
            raw_input=ToolInputBag(
                {
                    "child_session_id": child,
                    "subagent_id": child,
                    "status": "cancelled" if cancelled else "completed",
                    "duration_ms": data.get("durationMs"),
                    "tool_calls": data.get("totalToolCalls"),
                    "tokens_used": data.get("totalTokens"),
                }
            ),
            update_index=index,
        )
    return None


def _model_from_events(events: list[JsonObject]) -> str:
    for row in reversed(events):
        typ = str(row.get("type") or "")
        data = json_mapping(row.get("data"))
        if typ in {"assistant.message", "tool.execution_start", "session.shutdown"}:
            mid = str(data.get("model") or data.get("currentModel") or "").strip()
            if mid:
                return mid
    return ""


def _version_from_events(events: list[JsonObject]) -> str:
    for row in events:
        if str(row.get("type") or "") != "session.start":
            continue
        ver = str(json_mapping(row.get("data")).get("copilotVersion") or "").strip()
        if ver:
            return ver
    return ""


def _count_tools(events: list[JsonObject]) -> int:
    return sum(1 for row in events if str(row.get("type") or "") == "tool.execution_start")


def _count_subagents(events: list[JsonObject]) -> int:
    return sum(1 for row in events if str(row.get("type") or "") == "subagent.started")


def _meta_from_row(row: sqlite3.Row, db: Path, events: list[JsonObject]) -> SessionMeta:
    sid = str(row["id"])
    created = Stamp.iso(row["created_at"])
    updated = Stamp.iso(row["updated_at"] or row["created_at"])
    duration = 0.0
    start = Stamp.epoch(row["created_at"])
    end = Stamp.epoch(row["updated_at"] or row["created_at"])
    if start is not None and end is not None:
        duration = max(0.0, float(end - start))
    cwd = str(row["cwd"] or "").strip()
    kids = _count_subagents(events)
    return SessionMeta(
        session_id=sid,
        session_dir=db,
        model_id=_model_from_events(events) or "unknown",
        title=str(row["summary"] or "").strip(),
        created_at=created,
        updated_at=updated,
        duration_seconds=duration,
        run_dir=cwd,
        num_events=len(_timeline_for(events)),
        tool_call_count=_count_tools(events),
        turn_outcome=_turn_outcome(events),
        harness=COPILOT_HARNESS_ID,
        harness_version=_version_from_events(events),
        has_subagents=kids > 0,
        subagent_count=kids,
    )


class CopilotAdapter:
    """``~/.copilot`` sqlite catalog plus ``session-state`` event logs."""

    id = COPILOT_HARNESS_ID
    product = "GitHub Copilot"
    supported_version = "1.0.82"

    def db(self) -> Path:
        return default_db_path()

    def default_host_roots(self) -> list[Path]:
        path = self.db()
        return [path] if path.is_file() else []

    def discover(self, roots: Sequence[Path | str] | None = None) -> list[SessionRef]:
        found: list[SessionRef] = []
        seen: set[str] = set()
        for db in self._dbs_in(roots):
            for ref in self._discover_db(db):
                if ref.session_id in seen:
                    continue
                seen.add(ref.session_id)
                found.append(ref)
        return found

    def looks_like(self, ref: SessionRef | Path | str) -> bool:
        if isinstance(ref, SessionRef):
            return ref.harness == COPILOT_HARNESS_ID
        from .ref import parse_session_ref_string

        parsed = parse_session_ref_string(str(ref))
        if parsed is not None:
            return parsed[0] == COPILOT_HARNESS_ID
        path = Path(str(ref)).expanduser()
        return path.is_file() and path.name == "session-store.db"

    def bind_locator(self, locator: Path) -> SessionRef | None:
        """A database file is the store, not one session."""
        _ = Path(locator)
        return None

    def load_meta(self, ref: SessionRef | Path | str) -> SessionMeta:
        db, sid = _db_from_ref(ref, self.db())
        if not sid:
            raise FileNotFoundError("copilot session id is required")
        db = _assert_readable(db)
        with _connect(db) as con:
            row = _session_row(con, sid)
            if row is None:
                raise FileNotFoundError(f"copilot session not found: {sid}")
            events = list(json_lines(_events_path(db, sid)))
            return _meta_from_row(row, db, events)

    def parse_timeline(self, ref: SessionRef | Path | str) -> list[TraceEvent]:
        db, sid = _db_from_ref(ref, self.db())
        if not sid:
            return []
        db = _assert_readable(db)
        return _timeline_for(list(json_lines(_events_path(db, sid))))

    def ref_for_id(self, session_id: str) -> SessionRef | None:
        sid = (session_id or "").strip()
        if not sid:
            return None
        db = self.db()
        if not db.is_file():
            return None
        try:
            with _connect(db) as con:
                row = _session_row(con, sid)
        except sqlite3.Error:
            return None
        if row is None:
            return None
        return SessionRef(
            harness=COPILOT_HARNESS_ID,
            session_id=sid,
            locator=db,
            cwd=str(row["cwd"] or "").strip(),
        )

    def watch_hints(self) -> tuple[str, ...]:
        return ("session-store.db", "session-store.db-wal", "events.jsonl")

    def write_archive(self, ref: SessionRef | Path | str, dest: Path) -> list[str]:
        db, sid = _db_from_ref(ref, self.db())
        if not sid:
            raise RuntimeError("copilot session id is required")
        db = _assert_readable(db)
        dest = Path(dest)
        dest.parent.mkdir(parents=True, exist_ok=True)
        tmp = dest.with_name(dest.name + ".tmp")
        members: list[str] = []
        packed = False
        try:
            with tarfile.open(tmp, "w:gz") as tf:
                state = _state_dir(db, sid)
                for name in ("events.jsonl", "workspace.yaml", "session.db"):
                    src = state / name
                    if not src.is_file():
                        continue
                    arc = f"{sid}/{name}"
                    tf.add(src, arcname=arc)
                    members.append(arc)
            if not members:
                raise RuntimeError(f"copilot session has no archive files: {sid}")
            tmp.replace(dest)
            packed = True
        except (OSError, tarfile.TarError) as exc:
            raise RuntimeError(f"failed to pack session archive: {exc}") from exc
        finally:
            if not packed:
                tmp.unlink(missing_ok=True)
        return members

    def open_archive(self, src: Path, dest_root: Path) -> SessionRef:
        from .grok import extract_sid_tarball

        dest = extract_sid_tarball(src, dest_root)
        db = dest / "session.db"
        events = dest / "events.jsonl"
        if not db.is_file() and not events.is_file():
            raise RuntimeError(f"archive is not a copilot session: {src}")
        sid = dest.name
        store = dest_root / "copilot.db"
        if db.is_file():
            shutil.copy2(db, store)
        else:
            store.touch()
        state = _state_dir(store, sid)
        state.mkdir(parents=True, exist_ok=True)
        if events.is_file():
            shutil.copy2(events, state / "events.jsonl")
        workspace = dest / "workspace.yaml"
        if workspace.is_file():
            shutil.copy2(workspace, state / "workspace.yaml")
        return SessionRef(harness=COPILOT_HARNESS_ID, session_id=sid, locator=store)

    def load_detail(self, ref: SessionRef | Path | str) -> SessionMeta:
        return self.load_meta(ref)

    def timeline_stamp(self, ref: SessionRef | Path | str) -> tuple[float, int, int, int]:
        db, sid = _db_from_ref(ref, self.db())
        ev = Stamp.file(_events_path(db, sid)) if sid else (0.0, 0, 0, 0)
        store = Stamp.file(db)
        return (max(ev[0], store[0]), ev[1] + store[1], ev[2], ev[3])

    def trace_mtime(self, ref: SessionRef | Path | str) -> float:
        return self.timeline_stamp(ref)[0]

    def updates_size(self, ref: SessionRef | Path | str) -> int:
        return int(self.timeline_stamp(ref)[1])

    def scheduler_state(self, state: JsonObject) -> JsonObject | None:
        return None

    def reported_completion_ids(self, state: JsonObject) -> set[str]:
        return set()

    def list_turn_outcome(self, ref: SessionRef | Path | str) -> str:
        try:
            return (self.load_meta(ref).turn_outcome or "").strip()
        except FileNotFoundError:
            return ""

    def _dbs_in(self, roots: Sequence[Path | str] | None) -> list[Path]:
        if roots is None:
            path = self.db()
            return [path] if path.is_file() else []
        out: list[Path] = []
        for raw in roots:
            path = Path(raw).expanduser()
            if path.is_file() and path.name == "session-store.db":
                out.append(path)
            elif path.is_dir():
                cand = path / "session-store.db"
                if cand.is_file():
                    out.append(cand)
        return out

    def _discover_db(self, db: Path) -> list[SessionRef]:
        try:
            with _connect(db) as con:
                rows = _list_session_rows(con)
        except sqlite3.Error:
            return []
        found: list[SessionRef] = []
        for row in rows:
            found.append(
                SessionRef(
                    harness=COPILOT_HARNESS_ID,
                    session_id=str(row["id"]),
                    locator=db,
                    cwd=str(row["cwd"] or "").strip(),
                )
            )
        return found


__all__ = [
    "COPILOT_HARNESS_ID",
    "CopilotAdapter",
    "default_db_path",
    "default_store_root",
]

File: core/src/stores/copilot.rs
//! Copilot session-store.db plus events.jsonl.

use crate::event::{Event, EventType, ListMeta, ListStatus, SessionLocator};
use crate::jsonl;
use crate::store::Store;
use crate::text;
use rusqlite::Connection;
use serde_json::Value;
use std::path::{Path, PathBuf};

pub struct Copilot;

fn events_path(db: &Path, sid: &str) -> PathBuf {
    db.parent()
        .unwrap_or(db)
        .join("session-state")
        .join(sid)
        .join("events.jsonl")
}

fn from_row(row: &crate::jsonl::JsonlRow) -> Option<Event> {
    let typ = text::field_str(&row.value, "type");
    let data = row.value.get("data").cloned().unwrap_or(Value::Null);
    let ts = text::field_i64(&row.value, "timestamp");
    let eid = text::field_str(&row.value, "id");
    let ev = match typ.as_str() {
        "session.start" | "assistant.turn_start" => Event::new(EventType::TurnStarted)
            .with_ts(ts)
            .with_raw(&row.raw),
        "assistant.turn_end" => Event::new(EventType::TurnEnded)
            .with_ts(ts)
            .with_raw(&row.raw),
        "session.shutdown" => Event::new(EventType::TurnCompleted)
            .with_ts(ts)
            .with_content(text::field_str(&data, "shutdownType"))
            .with_raw(&row.raw),
        "user.message" => Event::new(EventType::UserMessageChunk)
            .with_ts(ts)
            .with_content(text::text_of(data.get("content").unwrap_or(&Value::Null)))
            .with_raw(&row.raw),
        "assistant.message" => Event::new(EventType::AgentMessageChunk)
            .with_ts(ts)
            .with_content(text::text_of(data.get("content").unwrap_or(&Value::Null)))
            .with_raw(&row.raw),
        "tool.execution_start" => {
            let args = data.get("arguments").cloned().unwrap_or(Value::Null);
            let mut ev = Event::new(EventType::ToolCall)
                .with_ts(ts)
                .with_raw(serde_json::to_string(&args).unwrap_or_else(|_| row.raw.clone()));
            ev.tool_name = text::field_str(&data, "toolName");
            ev.tool_call_id = {
                let id = text::field_str(&data, "toolCallId");
                if id.is_empty() {
                    eid
                } else {
                    id
                }
            };
            ev
        }
        "tool.execution_complete" => {
            let mut ev = Event::new(EventType::ToolCallUpdate)
                .with_ts(ts)
                .with_content(text::text_of(data.get("result").unwrap_or(&Value::Null)))
                .with_raw(&row.raw);
            ev.tool_call_id = text::field_str(&data, "toolCallId");
            ev.is_error = data.get("success") != Some(&Value::Bool(true));
            ev
        }
        "subagent.started" => {
            let mut ev = Event::new(EventType::SubagentSpawned)
                .with_ts(ts)
                .with_content(text::field_str(&data, "agentName"))
                .with_raw(&row.raw);
            ev.child_session_id = text::field_str(&row.value, "agentId");
            if ev.child_session_id.is_empty() {
                ev.child_session_id = text::field_str(&data, "agentId");
            }
            ev.subagent_type = text::field_str(&data, "agentType");
            ev.description = text::field_str(&data, "agentDescription");
            ev
        }
        "subagent.completed" => {
            let mut ev = Event::new(EventType::SubagentFinished)
                .with_ts(ts)
                .with_raw(&row.raw);
            ev.child_session_id = text::field_str(&row.value, "agentId");
            if ev.child_session_id.is_empty() {
                ev.child_session_id = text::field_str(&data, "agentId");
            }
            ev.subagent_type = text::field_str(&data, "agentType");
            ev
        }
        _ => return None,
    };
    Some(ev)
}

impl Store for Copilot {
    fn id(&self) -> &'static str {
        "copilot"
    }

    fn discover(&self, roots: &[PathBuf]) -> Vec<SessionLocator> {
        let mut out = Vec::new();
        for raw in roots {
            let db = if raw.is_file() {
                raw.clone()
            } else {
                raw.join("session-store.db")
            };
            if !db.is_file() {
                continue;
            }
            let Ok(con) =
                Connection::open_with_flags(&db, rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY)
            else {
                continue;
            };
            let Ok(mut stmt) = con.prepare("SELECT id, cwd FROM sessions") else {
                continue;
            };
            let rows = stmt.query_map([], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, String>(1).unwrap_or_default(),
                ))
            });
            let Ok(rows) = rows else { continue };
            for row in rows.flatten() {
                out.push(SessionLocator {
                    harness: "copilot".into(),
                    session_id: row.0,
                    locator: db.clone(),
                    cwd: row.1,
                });
            }
        }
        out
    }

    fn records(
        &self,
        locator: &Path,
        session_id: &str,
    ) -> Result<Vec<crate::store::Record>, String> {
        let path = if locator.extension().and_then(|s| s.to_str()) == Some("jsonl") {
            locator.to_path_buf()
        } else {
            events_path(locator, session_id)
        };
        if !path.is_file() {
            return Ok(Vec::new());
        }
        Ok(jsonl::cached_records(&path, None))
    }

    fn events(&self, records: &[crate::store::Record]) -> Vec<Event> {
        records.iter().filter_map(from_row).collect()
    }

    fn stamp(&self, locator: &Path, session_id: &str) -> crate::event::FileStamp {
        let path = if locator.extension().and_then(|s| s.to_str()) == Some("jsonl") {
            locator.to_path_buf()
        } else {
            events_path(locator, session_id)
        };
        jsonl::file_stamp(&path)
    }

    fn list_meta(&self, locator: &Path, session_id: &str) -> Result<ListMeta, String> {
        if session_id.is_empty() {
            return Err("copilot session id is required".into());
        }
        if !locator.is_file() {
            return Err(format!("copilot database not found: {}", locator.display()));
        }
        let con = Connection::open_with_flags(locator, rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY)
            .map_err(|e| e.to_string())?;
        let row = con
            .query_row(
                "SELECT id, cwd, summary, created_at, updated_at FROM sessions WHERE id = ?1",
                [session_id],
                |r| {
                    Ok((
                        r.get::<_, String>(0)?,
                        r.get::<_, String>(1).unwrap_or_default(),
                        r.get::<_, String>(2).unwrap_or_default(),
                        r.get::<_, rusqlite::types::Value>(3)?,
                        r.get::<_, rusqlite::types::Value>(4)?,
                    ))
                },
            )
            .map_err(|_| format!("copilot session not found: {session_id}"))?;
        let events = jsonl::window(&events_path(locator, session_id));
        Ok(meta_from_row(locator, &row, &events))
    }
}

const TURN_SIGNALS: &[&str] = &[
    "assistant.turn_start",
    "tool.execution_start",
    "subagent.started",
    "session.shutdown",
    "assistant.turn_end",
];

fn sql_stamp(val: &rusqlite::types::Value) -> String {
    match val {
        rusqlite::types::Value::Text(s) => text::iso_stamp(&Value::String(s.clone())),
        rusqlite::types::Value::Integer(n) => text::iso_millis(*n),
        rusqlite::types::Value::Real(n) => text::iso_millis(*n as i64),
        _ => String::new(),
    }
}

fn sql_epoch(val: &rusqlite::types::Value) -> Option<i64> {
    match val {
        rusqlite::types::Value::Text(s) => text::epoch_secs(&Value::String(s.clone())),
        rusqlite::types::Value::Integer(n) => text::epoch_secs(&Value::Number((*n).into())),
        rusqlite::types::Value::Real(n) => text::epoch_secs(&Value::from(*n)),
        _ => None,
    }
}

fn last_turn_type(events: &[crate::jsonl::JsonlRow]) -> String {
    let mut last = String::new();
    for ev in events {
        let typ = text::field_str(&ev.value, "type");
        if TURN_SIGNALS.contains(&typ.as_str()) {
            last = typ;
        }
    }
    last
}

fn model_from_events(events: &[crate::jsonl::JsonlRow]) -> String {
    for row in events.iter().rev() {
        let typ = text::field_str(&row.value, "type");
        let data = row.value.get("data").cloned().unwrap_or(Value::Null);
        if matches!(
            typ.as_str(),
            "assistant.message" | "tool.execution_start" | "session.shutdown"
        ) {
            let mid = text::field_str(&data, "model");
            if !mid.is_empty() {
                return mid;
            }
            let mid = text::field_str(&data, "currentModel");
            if !mid.is_empty() {
                return mid;
            }
        }
    }
    String::new()
}

fn version_from_events(events: &[crate::jsonl::JsonlRow]) -> String {
    for row in events {
        if text::field_str(&row.value, "type") != "session.start" {
            continue;
        }
        let ver = text::field_str(
            &row.value.get("data").cloned().unwrap_or(Value::Null),
            "copilotVersion",
        );
        if !ver.is_empty() {
            return ver;
        }
    }
    String::new()
}

fn count_type(events: &[crate::jsonl::JsonlRow], want: &str) -> u32 {
    events
        .iter()
        .filter(|row| text::field_str(&row.value, "type") == want)
        .count() as u32
}

fn meta_from_row(
    db: &Path,
    row: &(
        String,
        String,
        String,
        rusqlite::types::Value,
        rusqlite::types::Value,
    ),
    events: &[crate::jsonl::JsonlRow],
) -> ListMeta {
    let (sid, cwd, summary, created_raw, updated_raw) = row;
    let created = sql_stamp(created_raw);
    let mut updated = sql_stamp(updated_raw);
    if updated.is_empty() {
        updated = created.clone();
    }
    let kids = count_type(events, "subagent.started");
    let model = model_from_events(events);
    ListMeta {
        session_id: sid.clone(),
        locator: db.to_path_buf(),
        model_id: if model.is_empty() {
            "unknown".into()
        } else {
            model
        },
        title: summary.trim().to_string(),
        created_at: created,
        updated_at: updated,
        duration_seconds: text::duration_secs(sql_epoch(created_raw), sql_epoch(updated_raw)),
        tool_call_count: count_type(events, "tool.execution_start"),
        turn_outcome: {
            let mapped = ListStatus::from_token(&last_turn_type(events));
            if mapped != ListStatus::Idle {
                mapped.as_str().into()
            } else {
                String::new()
            }
        },
        harness: "copilot".into(),
        harness_version: version_from_events(events),
        run_dir: cwd.trim().to_string(),
        num_events: 0,
        has_subagents: kids > 0,
        subagent_count: kids,
        context_tokens_used: None,
    }
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::io::Write;
    use std::path::{Path, PathBuf};
    use std::time::{SystemTime, UNIX_EPOCH};

    fn temp_dir(label: &str) -> PathBuf {
        let dir = std::env::temp_dir().join(format!(
            "anqa-copilot-{label}-{}-{}",
            std::process::id(),
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_nanos()
        ));
        let _ = fs::remove_dir_all(&dir);
        fs::create_dir_all(&dir).unwrap();
        dir
    }

    fn write_events(root: &Path, sid: &str, line: &str) -> PathBuf {
        let path = root.join("session-state").join(sid).join("events.jsonl");
        fs::create_dir_all(path.parent().unwrap()).unwrap();
        fs::write(&path, format!("{line}\n")).unwrap();
        path
    }

    #[test]
    fn list_meta_window_title_and_last_turn() {
        let db = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("../tests/fixtures/harness/copilot/session-store.db");
        let meta =
            crate::list_meta("copilot", &db, "aaaaaaaa-1111-4111-8111-000000000001").unwrap();
        assert_eq!(meta.title, "Reply with COPILOT_PROBE_OK");
        assert_eq!(meta.turn_outcome, "complete");
        assert_eq!(meta.model_id, "gpt-5-mini");
    }

    #[test]
    fn copilot_stamp_follows_that_session_events_file() {
        let dir = temp_dir("stamp");
        let db = dir.join("session-store.db");
        fs::write(&db, b"shared").unwrap();
        write_events(&dir, "sid-a", r#"{"type":"user.message","id":"a1"}"#);
        write_events(
            &dir,
            "sid-b",
            r#"{"type":"user.message","id":"session-b-longer"}"#,
        );

        let stamp_a = crate::stamp("copilot", &db, "sid-a").unwrap();
        let stamp_b = crate::stamp("copilot", &db, "sid-b").unwrap();
        assert_ne!(stamp_a, stamp_b);

        let mut b = fs::OpenOptions::new()
            .append(true)
            .open(dir.join("session-state").join("sid-b").join("events.jsonl"))
            .unwrap();
        writeln!(b, r#"{{"type":"user.message","id":"b2"}}"#).unwrap();
        drop(b);

        assert_eq!(
            crate::stamp("copilot", &db, "sid-a").unwrap(),
            stamp_a,
            "growing B must not change A"
        );
        assert_ne!(crate::stamp("copilot", &db, "sid-b").unwrap(), stamp_b);

        fs::write(&db, b"shared-and-grown").unwrap();
        assert_eq!(
            crate::stamp("copilot", &db, "sid-a").unwrap(),
            stamp_a,
            "shared db must not be the stamp"
        );

        let mut a = fs::OpenOptions::new()
            .append(true)
            .open(dir.join("session-state").join("sid-a").join("events.jsonl"))
            .unwrap();
        writeln!(a, r#"{{"type":"user.message","id":"a2"}}"#).unwrap();
        drop(a);
        assert_ne!(crate::stamp("copilot", &db, "sid-a").unwrap(), stamp_a);

        let _ = fs::remove_dir_all(&dir);
    }
}
